src/main/java/com/google/devtools/build/lib/bazel/bzlmod/IndexRegistry.java

```diff
@@ -151,34 +172,58 @@ private <T> Optional<T> grabJson(String url, Class<T> klass, ExtendedEventHandle
     }
   }
 
-  @Override
-  public RepoSpec getRepoSpec(
-      ModuleKey key, RepositoryName repoName, ExtendedEventHandler eventHandler)
+  /**
+   * Grabs a JSON file from the given mdoule, and returns it as a parsed object with fields in {@code
+   * T}. Throws FileNotFoundException if the file does not exist.
+   */
+  private <T> T grabSourceJson(String url, ModuleKey key, Class<T> klass, ExtendedEventHandler eventHandler)
       throws IOException, InterruptedException {
-    Optional<SourceJson> sourceJson =
+    Optional<T> sourceJson =
         grabJson(
             constructUrl(
                 uri.toString(),
                 "modules",
                 key.getName(),
                 key.getVersion().toString(),
                 "source.json"),
-            SourceJson.class,
+            klass,
             eventHandler);
     if (sourceJson.isEmpty()) {
       throw new FileNotFoundException(
           String.format("Module %s's source information not found in registry %s", key, uri));
     }
+    return sourceJson.get();
+  }
 
-    String type = sourceJson.get().type;
-    switch (type) {
+  @Override
+  public RepoSpec getRepoSpec(
+      ModuleKey key, RepositoryName repoName, ExtendedEventHandler eventHandler)
+      throws IOException, InterruptedException {
+    String jsonUrl = constructUrl(
+        uri.toString(),
+        "modules",
+        key.getName(),
+        key.getVersion().toString(),
+        "source.json");
+    SourceJson sourceJson = grabSourceJson(jsonUrl, key, SourceJson.class, eventHandler);
+    switch (sourceJson.type) {
       case "archive":
-        return createArchiveRepoSpec(sourceJson, getBazelRegistryJson(eventHandler), key, repoName);
+        {
+          ArchiveSourceJson typedSourceJson = grabSourceJson(jsonUrl, key, ArchiveSourceJson.class, eventHandler);
```

Comment:
it looks like this actually downloads `jsonUrl` twice. Could you refactor the code so that `grabFile` is only called once per `getRepoSpec`?

Reply:
Done.